#### 7.1 异常概述与异常体系结构

- **java**程序在执行过程中所发生的异常事件可分为两类：

  > **<span style='color:red;'>Error</span>: Java虚拟机无法解决的严重问题：如：JVM系统内部错误、资源耗尽等严重情况。比如：<span style='color:red;'>StackOverflowError</span>和<span style='color:red;'>OOM</span>**

  > <span style='color:red;'>**Exception**</span>:其他因编程错误或偶然的外在因素导致的一般性问题，可以使用针对性的代码进行处理。例如：
  >
  > 1. 空指针访问
  > 2. 试图读取不存在的文件
  > 3. 网络连接中断
  > 4. 数组角标越界
  
  分为编译时异常和运行时异常
  
  ![image-20200913134527843](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913134527843.png)

#### 7.2 常见异常

![image-20200913134543557](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913134543557.png)

#### 7.3 异常处理机制

**java采用的异常处理机制，是将异常处理的程序代码集中在一起，与正常的程序代码分开，使得程序简洁、优雅，并易于维护。**

**方式一：try - catch - finally**

**方式二：throws + 异常类型**

![image-20200913135321451](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913135321451.png)

##### 7.3.1 try-catch-finally

**如果一个方法内抛出异常， 该异常对象会被抛给调用者方法中处理。 如果异常没有在调用者方法中处理， 它继续被抛给这个调用方法的上层方法。 这个过程将一直继续下去， 直到异常被处理。这一过程称为**<span style='color:red;'>**捕获(catch)异常  **</span>

![image-20200913160137287](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913160137287.png)

如果一个异常回到main()方法，并且main()也不处理，则程序运行终止。

异常处理是通过try-catch-finally语句实现的。

![image-20200913161402758](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913161402758.png)

![image-20200913163142928](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913163142928.png)

![image-20200913163152586](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913163152586.png)

#### 不捕获异常时的情况

- 前面使用的异常都是RuntimeException类或是它的子类，这些类的异常的特点是：即使没有使用try和catch捕获， Java自己也能捕获，并且编译通过( 但运行时会发生异常使得程序运行终止 )。

- 如果抛出的异常是IOException等类型的非运行时异常，则必须捕获，否则编译错误。也就是说，我们必须处理编译时异常，将异常进行捕捉，转化为运行时异常  

#### 7.4 异常处理机制二：声明抛出异常

- 如果一个方法(中的语句执行时)可能生成某种异常， 但是并不能确定如何处理这种异常， 则此方法应<span style='color:red;'>**显示地**</span>声明抛出异常， 表明该方法将不对这些异常进行处理，而由该方法的<span style='color:red;'>**调用者**</span>负责处理。  
- 在方法声明中用<span style='color:red;'>**throws**</span>语句可以声明抛出异常的列表，throws后面的异常类型可以是方法中产生的异常类型，也可以是它的父类。
- ![image-20200913202657942](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913202657942.png)

重写方法不能抛出比被重写方法范围更大的异常类型。在多态的情况下，对methodA()方法的调用-异常的捕获按父类声明的异常处理。![image-20200913211328242](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913211328242.png)

#### 7.5 手动抛出异常

- Java异常类对象除在程序执行过程中出现异常时由系统自动生成并
  抛出， 也可根据需要使用人工创建并抛出。
- ![image-20200913211446394](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913211446394.png)

#### 7.6 用户自定义异常类

- 一般地，用户自定义异常类都是<span style='color:red;'>**RuntimeException**</span>的子类。

- 自定义异常类通常需要编写几个<span style='color:red;'>**重载的构造器**</span>。
- 自定义异常需要提供<span style='color:red;'>**serialVersionUID**</span>
-  自定义的异常通过<span style='color:red;'>**throw**</span>抛出。
-  自定义异常最重要的是异常类的名字，当异常出现时，可以根据名字判断异常类型。  

用户自己的异常类必须继承现有的异常类。

![image-20200913195316544](C:\Users\chenlin\AppData\Roaming\Typora\typora-user-images\image-20200913195316544.png)